### PR TITLE
Optimize memory usage in Assemly::init()

### LINKED ISSUE #16228: Optimize memory usage in Assembly::init()
## Reason

When there are many variables on each element, some members in Assembly uses a lot of memory because we consider full coupling by default.

Create memory spikes.

```
(pprof) list Assembly::init
Total: 1004.79MB
ROUTINE ======================== Assembly::init in framework/src/base/Assembly.C
     0  637.57MB (flat, cum) 63.45% of Total
     .     .  2373: _block_diagonal_matrix = true;
     .     .  2374: for (auto & ivar : vars)
     .     .  2375: {
     .     .  2376:  auto i = ivar->number();
     .     .  2377:  if (i >= _component_block_diagonal.size())
     .   1.01kB  2378:   _component_block_diagonal.resize(i + 1, true);
     .     .  2379:
     .     .  2380:  auto ivar_start = _cm_ff_entry.size();
     .     .  2381:  for (unsigned int k = 0; k < ivar->count(); ++k)
     .     .  2382:  {
     .     .  2383:   unsigned int iv = i + k;
     .     .  2384:   for (const auto & j : ConstCouplingRow(iv, *_cm))
     .     .  2385:   {
     .     .  2386:    if (_sys.isScalarVariable(j))
     .     .  2387:    {
     .     .  2388:     auto & jvar = _sys.getScalarVariable(_tid, j);
     .     .  2389:     _cm_fs_entry.push_back(std::make_pair(ivar, &jvar));
     .     .  2390:     _block_diagonal_matrix = false;
     .     .  2391:    }
     .     .  2392:    else
     .     .  2393:    {
     .     .  2394:     auto & jvar = _sys.getVariable(_tid, j);
     .     .  2395:     auto pair = std::make_pair(ivar, &jvar);
     .     .  2396:     auto c = ivar_start;
     .     .  2397:     // check if the pair has been pushed or not
     .     .  2398:     bool has_pair = false;
     .     .  2399:     for (; c < _cm_ff_entry.size(); ++c)
     .     .  2400:      if (_cm_ff_entry[c] == pair)
     .     .  2401:      {
     .     .  2402:       has_pair = true;
     .     .  2403:       break;
     .     .  2404:      }
     .     .  2405:     if (!has_pair)
     .     .  2406:      _cm_ff_entry.push_back(pair);
     .     .  2407:     // only set having diagonal matrix to false when ivar and jvar numbers are different
     .     .  2408:     // Note: for array variables, since we save the entire local Jacobian of all components,
     .     .  2409:     //    even there are couplings among components of the same array variable, we still
     .     .  2410:     //    do not set the flag to false.
     .     .  2411:     if (i != jvar.number())
     .     .  2412:      _block_diagonal_matrix = false;
     .     .  2413:     else if (iv != j)
     .     .  2414:      _component_block_diagonal[i] = false;
     .     .  2415:    }
     .     .  2416:   }
     .     .  2417:  }
     .     .  2418: }
     .     .  2419:
     .     .  2420: auto & scalar_vars = _sys.getScalarVariables(_tid);
     .     .  2421:
     .     .  2422: for (auto & ivar : scalar_vars)
     .     .  2423: {
     .     .  2424:  auto i = ivar->number();
     .     .  2425:  if (i >= _component_block_diagonal.size())
     .     .  2426:   _component_block_diagonal.resize(i + 1, true);
     .     .  2427:
     .     .  2428:  for (const auto & j : ConstCouplingRow(i, *_cm))
     .     .  2429:   if (_sys.isScalarVariable(j))
     .     .  2430:   {
     .     .  2431:    auto & jvar = _sys.getScalarVariable(_tid, j);
     .     .  2432:    _cm_ss_entry.push_back(std::make_pair(ivar, &jvar));
     .     .  2433:   }
     .     .  2434:   else
     .     .  2435:   {
     .     .  2436:    auto & jvar = _sys.getVariable(_tid, j);
     .     .  2437:    _cm_sf_entry.push_back(std::make_pair(ivar, &jvar));
     .     .  2438:   }
     .     .  2439: }
     .     .  2440:
     .     .  2441: if (_block_diagonal_matrix && scalar_vars.size() != 0)
     .     .  2442:  _block_diagonal_matrix = false;
     .     .  2443:
     .     .  2444: auto num_vector_tags = _residual_vector_tags.size();
     .     .  2445:
     .     .  2446: _sub_Re.resize(num_vector_tags);
     .     .  2447: _sub_Rn.resize(num_vector_tags);
     .     .  2448: _sub_Rl.resize(num_vector_tags);
     .     .  2449: for (MooseIndex(_sub_Re) i = 0; i < _sub_Re.size(); i++)
     .     .  2450: {
     .   704kB  2451:  _sub_Re[i].resize(n_vars);
     .   704kB  2452:  _sub_Rn[i].resize(n_vars);
     .   704kB  2453:  _sub_Rl[i].resize(n_vars);
     .     .  2454: }
     .     .  2455:
     .     .  2456: _cached_residual_values.resize(num_vector_tags);
     .     .  2457: _cached_residual_rows.resize(num_vector_tags);
     .     .  2458:
     .     .  2459: auto num_matrix_tags = _subproblem.numMatrixTags();
     .     .  2460:
     .     .  2461: _cached_jacobian_values.resize(num_matrix_tags);
     .     .  2462: _cached_jacobian_rows.resize(num_matrix_tags);
     .     .  2463: _cached_jacobian_cols.resize(num_matrix_tags);
     .     .  2464:
     .     .  2465: // Element matrices
     .     .  2466: _sub_Kee.resize(num_matrix_tags);
     .     .  2467: _sub_Keg.resize(num_matrix_tags);
     .     .  2468: _sub_Ken.resize(num_matrix_tags);
     .     .  2469: _sub_Kne.resize(num_matrix_tags);
     .     .  2470: _sub_Knn.resize(num_matrix_tags);
     .     .  2471: _sub_Kll.resize(num_matrix_tags);
     .     .  2472: _sub_Kle.resize(num_matrix_tags);
     .     .  2473: _sub_Kln.resize(num_matrix_tags);
     .     .  2474: _sub_Kel.resize(num_matrix_tags);
     .     .  2475: _sub_Knl.resize(num_matrix_tags);
     .     .  2476:
     .     .  2477: _jacobian_block_used.resize(num_matrix_tags);
     .     .  2478: _jacobian_block_neighbor_used.resize(num_matrix_tags);
     .     .  2479: _jacobian_block_lower_used.resize(num_matrix_tags);
     .     .  2480: _jacobian_block_nonlocal_used.resize(num_matrix_tags);
     .     .  2481:
     .     .  2482: for (MooseIndex(num_matrix_tags) tag = 0; tag < num_matrix_tags; tag++)
     .     .  2483: {
     .   660kB  2484:  _sub_Keg[tag].resize(n_vars);
     .   660kB  2485:  _sub_Ken[tag].resize(n_vars);
     .   660kB  2486:  _sub_Kne[tag].resize(n_vars);
     .   660kB  2487:  _sub_Knn[tag].resize(n_vars);
     .   660kB  2488:  _sub_Kee[tag].resize(n_vars);
     .   660kB  2489:  _sub_Kll[tag].resize(n_vars);
     .   660kB  2490:  _sub_Kle[tag].resize(n_vars);
     .   660kB  2491:  _sub_Kln[tag].resize(n_vars);
     .   660kB  2492:  _sub_Kel[tag].resize(n_vars);
     .   660kB  2493:  _sub_Knl[tag].resize(n_vars);
     .     .  2494:
     .   660kB  2495:  _jacobian_block_used[tag].resize(n_vars);
     .   660kB  2496:  _jacobian_block_neighbor_used[tag].resize(n_vars);
     .   660kB  2497:  _jacobian_block_lower_used[tag].resize(n_vars);
     .   660kB  2498:  _jacobian_block_nonlocal_used[tag].resize(n_vars);
     .     .  2499:  for (MooseIndex(n_vars) i = 0; i < n_vars; ++i)
     .     .  2500:  {
     .     .  2501:   if (!_block_diagonal_matrix)
     .     .  2502:   {
     .     .  2503:    _sub_Kee[tag][i].resize(n_vars);
     .     .  2504:    _sub_Keg[tag][i].resize(n_vars);
     .     .  2505:    _sub_Ken[tag][i].resize(n_vars);
     .     .  2506:    _sub_Kne[tag][i].resize(n_vars);
     .     .  2507:    _sub_Knn[tag][i].resize(n_vars);
     .     .  2508:    _sub_Kll[tag][i].resize(n_vars);
     .     .  2509:    _sub_Kle[tag][i].resize(n_vars);
     .     .  2510:    _sub_Kln[tag][i].resize(n_vars);
     .     .  2511:    _sub_Kel[tag][i].resize(n_vars);
     .     .  2512:    _sub_Knl[tag][i].resize(n_vars);
     .     .  2513:   }
     .     .  2514:   else
     .     .  2515:   {
     .   2.15MB  2516:    _sub_Kee[tag][i].resize(1);
     .   2.15MB  2517:    _sub_Keg[tag][i].resize(1);
     .   2.15MB  2518:    _sub_Ken[tag][i].resize(1);
     .   2.15MB  2519:    _sub_Kne[tag][i].resize(1);
     .   2.15MB  2520:    _sub_Knn[tag][i].resize(1);
     .   2.15MB  2521:    _sub_Kll[tag][i].resize(1);
     .   2.15MB  2522:    _sub_Kle[tag][i].resize(1);
     .   2.15MB  2523:    _sub_Kln[tag][i].resize(1);
     .   2.15MB  2524:    _sub_Kel[tag][i].resize(1);
     .   2.15MB  2525:    _sub_Knl[tag][i].resize(1);
     .     .  2526:   }
     .  151.25MB  2527:   _jacobian_block_used[tag][i].resize(n_vars);
     .  151.25MB  2528:   _jacobian_block_neighbor_used[tag][i].resize(n_vars);
     .  151.25MB  2529:   _jacobian_block_lower_used[tag][i].resize(n_vars);
     .  151.25MB  2530:   _jacobian_block_nonlocal_used[tag][i].resize(n_vars);
     .     .  2531:  }
     .     .  2532: }
     .     .  2533:}
     .     .  2534:
     .     .  2535:void
```

## Design

If there is only diagonal coupling, we do not allocate memory for all variables. 

## Impact
Improve memory efficiency for problems with many moose variables.

### PR DESCRIPTION
**Pull request by fdkong**
Close #16228

### CODE REVIEW

**Comment by YaqiWang** [comment]
LGTM!

**Comment by YaqiWang** [comment]
Just to make sure: Fande, did you try this with the modified test I mentioned?

**Comment by fdkong** [comment]
Before this PR:

```
(pprof) list Assembly::init
Total: 1004.79MB
ROUTINE ======================== Assembly::init in framework/src/base/Assembly.C
     0  637.57MB (flat, cum) 63.45% of Total
     .     .  2373: _block_diagonal_matrix = true;
     .     .  2374: for (auto & ivar : vars)
     .     .  2375: {
     .     .  2376:  auto i = ivar->number();
     .     .  2377:  if (i >= _component_block_diagonal.size())
     .   1.01kB  2378:   _component_block_diagonal.resize(i + 1, true);
     .     .  2379:
     .     .  2380:  auto ivar_start = _cm_ff_entry.size();
     .     .  2381:  for (unsigned int k = 0; k < ivar->count(); ++k)
     .     .  2382:  {
     .     .  2383:   unsigned int iv = i + k;
     .     .  2384:   for (const auto & j : ConstCouplingRow(iv, *_cm))
     .     .  2385:   {
     .     .  2386:    if (_sys.isScalarVariable(j))
     .     .  2387:    {
     .     .  2388:     auto & jvar = _sys.getScalarVariable(_tid, j);
     .     .  2389:     _cm_fs_entry.push_back(std::make_pair(ivar, &jvar));
     .     .  2390:     _block_diagonal_matrix = false;
     .     .  2391:    }
     .     .  2392:    else
     .     .  2393:    {
     .     .  2394:     auto & jvar = _sys.getVariable(_tid, j);
     .     .  2395:     auto pair = std::make_pair(ivar, &jvar);
     .     .  2396:     auto c = ivar_start;
     .     .  2397:     // check if the pair has been pushed or not
     .     .  2398:     bool has_pair = false;
     .     .  2399:     for (; c < _cm_ff_entry.size(); ++c)
     .     .  2400:      if (_cm_ff_entry[c] == pair)
     .     .  2401:      {
     .     .  2402:       has_pair = true;
     .     .  2403:       break;
     .     .  2404:      }
     .     .  2405:     if (!has_pair)
     .     .  2406:      _cm_ff_entry.push_back(pair);
     .     .  2407:     // only set having diagonal matrix to false when ivar and jvar numbers are different
     .     .  2408:     // Note: for array variables, since we save the entire local Jacobian of all components,
     .     .  2409:     //    even there are couplings among components of the same array variable, we still
     .     .  2410:     //    do not set the flag to false.
     .     .  2411:     if (i != jvar.number())
     .     .  2412:      _block_diagonal_matrix = false;
     .     .  2413:     else if (iv != j)
     .     .  2414:      _component_block_diagonal[i] = false;
     .     .  2415:    }
     .     .  2416:   }
     .     .  2417:  }
     .     .  2418: }
     .     .  2419:
     .     .  2420: auto & scalar_vars = _sys.getScalarVariables(_tid);
     .     .  2421:
     .     .  2422: for (auto & ivar : scalar_vars)
     .     .  2423: {
     .     .  2424:  auto i = ivar->number();
     .     .  2425:  if (i >= _component_block_diagonal.size())
     .     .  2426:   _component_block_diagonal.resize(i + 1, true);
     .     .  2427:
     .     .  2428:  for (const auto & j : ConstCouplingRow(i, *_cm))
     .     .  2429:   if (_sys.isScalarVariable(j))
     .     .  2430:   {
     .     .  2431:    auto & jvar = _sys.getScalarVariable(_tid, j);
     .     .  2432:    _cm_ss_entry.push_back(std::make_pair(ivar, &jvar));
     .     .  2433:   }
     .     .  2434:   else
     .     .  2435:   {
     .     .  2436:    auto & jvar = _sys.getVariable(_tid, j);
     .     .  2437:    _cm_sf_entry.push_back(std::make_pair(ivar, &jvar));
     .     .  2438:   }
     .     .  2439: }
     .     .  2440:
     .     .  2441: if (_block_diagonal_matrix && scalar_vars.size() != 0)
     .     .  2442:  _block_diagonal_matrix = false;
     .     .  2443:
     .     .  2444: auto num_vector_tags = _residual_vector_tags.size();
     .     .  2445:
     .     .  2446: _sub_Re.resize(num_vector_tags);
     .     .  2447: _sub_Rn.resize(num_vector_tags);
     .     .  2448: _sub_Rl.resize(num_vector_tags);
     .     .  2449: for (MooseIndex(_sub_Re) i = 0; i < _sub_Re.size(); i++)
     .     .  2450: {
     .   704kB  2451:  _sub_Re[i].resize(n_vars);
     .   704kB  2452:  _sub_Rn[i].resize(n_vars);
     .   704kB  2453:  _sub_Rl[i].resize(n_vars);
     .     .  2454: }
     .     .  2455:
     .     .  2456: _cached_residual_values.resize(num_vector_tags);
     .     .  2457: _cached_residual_rows.resize(num_vector_tags);
     .     .  2458:
     .     .  2459: auto num_matrix_tags = _subproblem.numMatrixTags();
     .     .  2460:
     .     .  2461: _cached_jacobian_values.resize(num_matrix_tags);
     .     .  2462: _cached_jacobian_rows.resize(num_matrix_tags);
     .     .  2463: _cached_jacobian_cols.resize(num_matrix_tags);
     .     .  2464:
     .     .  2465: // Element matrices
     .     .  2466: _sub_Kee.resize(num_matrix_tags);
     .     .  2467: _sub_Keg.resize(num_matrix_tags);
     .     .  2468: _sub_Ken.resize(num_matrix_tags);
     .     .  2469: _sub_Kne.resize(num_matrix_tags);
     .     .  2470: _sub_Knn.resize(num_matrix_tags);
     .     .  2471: _sub_Kll.resize(num_matrix_tags);
     .     .  2472: _sub_Kle.resize(num_matrix_tags);
     .     .  2473: _sub_Kln.resize(num_matrix_tags);
     .     .  2474: _sub_Kel.resize(num_matrix_tags);
     .     .  2475: _sub_Knl.resize(num_matrix_tags);
     .     .  2476:
     .     .  2477: _jacobian_block_used.resize(num_matrix_tags);
     .     .  2478: _jacobian_block_neighbor_used.resize(num_matrix_tags);
     .     .  2479: _jacobian_block_lower_used.resize(num_matrix_tags);
     .     .  2480: _jacobian_block_nonlocal_used.resize(num_matrix_tags);
     .     .  2481:
     .     .  2482: for (MooseIndex(num_matrix_tags) tag = 0; tag < num_matrix_tags; tag++)
     .     .  2483: {
     .   660kB  2484:  _sub_Keg[tag].resize(n_vars);
     .   660kB  2485:  _sub_Ken[tag].resize(n_vars);
     .   660kB  2486:  _sub_Kne[tag].resize(n_vars);
     .   660kB  2487:  _sub_Knn[tag].resize(n_vars);
     .   660kB  2488:  _sub_Kee[tag].resize(n_vars);
     .   660kB  2489:  _sub_Kll[tag].resize(n_vars);
     .   660kB  2490:  _sub_Kle[tag].resize(n_vars);
     .   660kB  2491:  _sub_Kln[tag].resize(n_vars);
     .   660kB  2492:  _sub_Kel[tag].resize(n_vars);
     .   660kB  2493:  _sub_Knl[tag].resize(n_vars);
     .     .  2494:
     .   660kB  2495:  _jacobian_block_used[tag].resize(n_vars);
     .   660kB  2496:  _jacobian_block_neighbor_used[tag].resize(n_vars);
     .   660kB  2497:  _jacobian_block_lower_used[tag].resize(n_vars);
     .   660kB  2498:  _jacobian_block_nonlocal_used[tag].resize(n_vars);
     .     .  2499:  for (MooseIndex(n_vars) i = 0; i < n_vars; ++i)
     .     .  2500:  {
     .     .  2501:   if (!_block_diagonal_matrix)
     .     .  2502:   {
     .     .  2503:    _sub_Kee[tag][i].resize(n_vars);
     .     .  2504:    _sub_Keg[tag][i].resize(n_vars);
     .     .  2505:    _sub_Ken[tag][i].resize(n_vars);
     .     .  2506:    _sub_Kne[tag][i].resize(n_vars);
     .     .  2507:    _sub_Knn[tag][i].resize(n_vars);
     .     .  2508:    _sub_Kll[tag][i].resize(n_vars);
     .     .  2509:    _sub_Kle[tag][i].resize(n_vars);
     .     .  2510:    _sub_Kln[tag][i].resize(n_vars);
     .     .  2511:    _sub_Kel[tag][i].resize(n_vars);
     .     .  2512:    _sub_Knl[tag][i].resize(n_vars);
     .     .  2513:   }
     .     .  2514:   else
     .     .  2515:   {
     .   2.15MB  2516:    _sub_Kee[tag][i].resize(1);
     .   2.15MB  2517:    _sub_Keg[tag][i].resize(1);
     .   2.15MB  2518:    _sub_Ken[tag][i].resize(1);
     .   2.15MB  2519:    _sub_Kne[tag][i].resize(1);
     .   2.15MB  2520:    _sub_Knn[tag][i].resize(1);
     .   2.15MB  2521:    _sub_Kll[tag][i].resize(1);
     .   2.15MB  2522:    _sub_Kle[tag][i].resize(1);
     .   2.15MB  2523:    _sub_Kln[tag][i].resize(1);
     .   2.15MB  2524:    _sub_Kel[tag][i].resize(1);
     .   2.15MB  2525:    _sub_Knl[tag][i].resize(1);
     .     .  2526:   }
     .  151.25MB  2527:   _jacobian_block_used[tag][i].resize(n_vars);
     .  151.25MB  2528:   _jacobian_block_neighbor_used[tag][i].resize(n_vars);
     .  151.25MB  2529:   _jacobian_block_lower_used[tag][i].resize(n_vars);
     .  151.25MB  2530:   _jacobian_block_nonlocal_used[tag][i].resize(n_vars);
     .     .  2531:  }
     .     .  2532: }
     .     .  2533:}
     .     .  2534:
     .     .  2535:void
```

After this PR:

```
(pprof) list Assembly::init
Total: 418.87MB
ROUTINE ======================== Assembly::init in framework/src/base/Assembly.C
         0    21.85MB (flat, cum)  5.22% of Total
         .          .   2373:  _block_diagonal_matrix = true;
         .          .   2374:  for (auto & ivar : vars)
         .          .   2375:  {
         .          .   2376:    auto i = ivar->number();
         .          .   2377:    if (i >= _component_block_diagonal.size())
         .     1.01kB   2378:      _component_block_diagonal.resize(i + 1, true);
         .          .   2379:
         .          .   2380:    auto ivar_start = _cm_ff_entry.size();
         .          .   2381:    for (unsigned int k = 0; k < ivar->count(); ++k)
         .          .   2382:    {
         .          .   2383:      unsigned int iv = i + k;
         .          .   2384:      for (const auto & j : ConstCouplingRow(iv, *_cm))
         .          .   2385:      {
         .          .   2386:        if (_sys.isScalarVariable(j))
         .          .   2387:        {
         .          .   2388:          auto & jvar = _sys.getScalarVariable(_tid, j);
         .          .   2389:          _cm_fs_entry.push_back(std::make_pair(ivar, &jvar));
         .          .   2390:          _block_diagonal_matrix = false;
         .          .   2391:        }
         .          .   2392:        else
         .          .   2393:        {
         .          .   2394:          auto & jvar = _sys.getVariable(_tid, j);
         .          .   2395:          auto pair = std::make_pair(ivar, &jvar);
         .          .   2396:          auto c = ivar_start;
         .          .   2397:          // check if the pair has been pushed or not
         .          .   2398:          bool has_pair = false;
         .          .   2399:          for (; c < _cm_ff_entry.size(); ++c)
         .          .   2400:            if (_cm_ff_entry[c] == pair)
         .          .   2401:            {
         .          .   2402:              has_pair = true;
         .          .   2403:              break;
         .          .   2404:            }
         .          .   2405:          if (!has_pair)
         .          .   2406:            _cm_ff_entry.push_back(pair);
         .          .   2407:          // only set having diagonal matrix to false when ivar and jvar numbers are different
         .          .   2408:          // Note: for array variables, since we save the entire local Jacobian of all components,
         .          .   2409:          //       even there are couplings among components of the same array variable, we still
         .          .   2410:          //       do not set the flag to false.
         .          .   2411:          if (i != jvar.number())
         .          .   2412:            _block_diagonal_matrix = false;
         .          .   2413:          else if (iv != j)
         .          .   2414:            _component_block_diagonal[i] = false;
         .          .   2415:        }
         .          .   2416:      }
         .          .   2417:    }
         .          .   2418:  }
         .          .   2419:
         .          .   2420:  auto & scalar_vars = _sys.getScalarVariables(_tid);
         .          .   2421:
         .          .   2422:  for (auto & ivar : scalar_vars)
         .          .   2423:  {
         .          .   2424:    auto i = ivar->number();
         .          .   2425:    if (i >= _component_block_diagonal.size())
         .          .   2426:      _component_block_diagonal.resize(i + 1, true);
         .          .   2427:
         .          .   2428:    for (const auto & j : ConstCouplingRow(i, *_cm))
         .          .   2429:      if (_sys.isScalarVariable(j))
         .          .   2430:      {
         .          .   2431:        auto & jvar = _sys.getScalarVariable(_tid, j);
         .          .   2432:        _cm_ss_entry.push_back(std::make_pair(ivar, &jvar));
         .          .   2433:      }
         .          .   2434:      else
         .          .   2435:      {
         .          .   2436:        auto & jvar = _sys.getVariable(_tid, j);
         .          .   2437:        _cm_sf_entry.push_back(std::make_pair(ivar, &jvar));
         .          .   2438:      }
         .          .   2439:  }
         .          .   2440:
         .          .   2441:  if (_block_diagonal_matrix && scalar_vars.size() != 0)
         .          .   2442:    _block_diagonal_matrix = false;
         .          .   2443:
         .          .   2444:  auto num_vector_tags = _residual_vector_tags.size();
         .          .   2445:
         .          .   2446:  _sub_Re.resize(num_vector_tags);
         .          .   2447:  _sub_Rn.resize(num_vector_tags);
         .          .   2448:  _sub_Rl.resize(num_vector_tags);
         .          .   2449:  for (MooseIndex(_sub_Re) i = 0; i < _sub_Re.size(); i++)
         .          .   2450:  {
         .      704kB   2451:    _sub_Re[i].resize(n_vars);
         .      704kB   2452:    _sub_Rn[i].resize(n_vars);
         .      704kB   2453:    _sub_Rl[i].resize(n_vars);
         .          .   2454:  }
         .          .   2455:
         .          .   2456:  _cached_residual_values.resize(num_vector_tags);
         .          .   2457:  _cached_residual_rows.resize(num_vector_tags);
         .          .   2458:
         .          .   2459:  auto num_matrix_tags = _subproblem.numMatrixTags();
         .          .   2460:
         .          .   2461:  _cached_jacobian_values.resize(num_matrix_tags);
         .          .   2462:  _cached_jacobian_rows.resize(num_matrix_tags);
         .          .   2463:  _cached_jacobian_cols.resize(num_matrix_tags);
         .          .   2464:
         .          .   2465:  // Element matrices
         .          .   2466:  _sub_Kee.resize(num_matrix_tags);
         .          .   2467:  _sub_Keg.resize(num_matrix_tags);
         .          .   2468:  _sub_Ken.resize(num_matrix_tags);
         .          .   2469:  _sub_Kne.resize(num_matrix_tags);
         .          .   2470:  _sub_Knn.resize(num_matrix_tags);
         .          .   2471:  _sub_Kll.resize(num_matrix_tags);
         .          .   2472:  _sub_Kle.resize(num_matrix_tags);
         .          .   2473:  _sub_Kln.resize(num_matrix_tags);
         .          .   2474:  _sub_Kel.resize(num_matrix_tags);
         .          .   2475:  _sub_Knl.resize(num_matrix_tags);
         .          .   2476:
         .          .   2477:  _jacobian_block_used.resize(num_matrix_tags);
         .          .   2478:  _jacobian_block_neighbor_used.resize(num_matrix_tags);
         .          .   2479:  _jacobian_block_lower_used.resize(num_matrix_tags);
         .          .   2480:  _jacobian_block_nonlocal_used.resize(num_matrix_tags);
         .          .   2481:
         .          .   2482:  for (MooseIndex(num_matrix_tags) tag = 0; tag < num_matrix_tags; tag++)
         .          .   2483:  {
         .      528kB   2484:    _sub_Keg[tag].resize(n_vars);
         .      528kB   2485:    _sub_Ken[tag].resize(n_vars);
         .      528kB   2486:    _sub_Kne[tag].resize(n_vars);
         .      528kB   2487:    _sub_Knn[tag].resize(n_vars);
         .      528kB   2488:    _sub_Kee[tag].resize(n_vars);
         .      528kB   2489:    _sub_Kll[tag].resize(n_vars);
         .      528kB   2490:    _sub_Kle[tag].resize(n_vars);
         .      528kB   2491:    _sub_Kln[tag].resize(n_vars);
         .      528kB   2492:    _sub_Kel[tag].resize(n_vars);
         .      528kB   2493:    _sub_Knl[tag].resize(n_vars);
         .          .   2494:
         .      528kB   2495:    _jacobian_block_used[tag].resize(n_vars);
         .      396kB   2496:    _jacobian_block_neighbor_used[tag].resize(n_vars);
         .      396kB   2497:    _jacobian_block_lower_used[tag].resize(n_vars);
         .      396kB   2498:    _jacobian_block_nonlocal_used[tag].resize(n_vars);
         .          .   2499:    for (MooseIndex(n_vars) i = 0; i < n_vars; ++i)
         .          .   2500:    {
         .          .   2501:      if (!_block_diagonal_matrix)
         .          .   2502:      {
         .          .   2503:        _sub_Kee[tag][i].resize(n_vars);
         .          .   2504:        _sub_Keg[tag][i].resize(n_vars);
         .          .   2505:        _sub_Ken[tag][i].resize(n_vars);
         .          .   2506:        _sub_Kne[tag][i].resize(n_vars);
         .          .   2507:        _sub_Knn[tag][i].resize(n_vars);
         .          .   2508:        _sub_Kll[tag][i].resize(n_vars);
         .          .   2509:        _sub_Kle[tag][i].resize(n_vars);
         .          .   2510:        _sub_Kln[tag][i].resize(n_vars);
         .          .   2511:        _sub_Kel[tag][i].resize(n_vars);
         .          .   2512:        _sub_Knl[tag][i].resize(n_vars);
         .          .   2513:
         .          .   2514:        _jacobian_block_used[tag][i].resize(n_vars);
         .          .   2515:        _jacobian_block_neighbor_used[tag][i].resize(n_vars);
         .          .   2516:        _jacobian_block_lower_used[tag][i].resize(n_vars);
         .          .   2517:        _jacobian_block_nonlocal_used[tag][i].resize(n_vars);
         .          .   2518:      }
         .          .   2519:      else
         .          .   2520:      {
         .     1.29MB   2521:        _sub_Kee[tag][i].resize(1);
         .     1.29MB   2522:        _sub_Keg[tag][i].resize(1);
         .     1.29MB   2523:        _sub_Ken[tag][i].resize(1);
         .     1.29MB   2524:        _sub_Kne[tag][i].resize(1);
         .     1.29MB   2525:        _sub_Knn[tag][i].resize(1);
         .     1.29MB   2526:        _sub_Kll[tag][i].resize(1);
         .     1.29MB   2527:        _sub_Kle[tag][i].resize(1);
         .     1.29MB   2528:        _sub_Kln[tag][i].resize(1);
         .     1.29MB   2529:        _sub_Kel[tag][i].resize(1);
         .     1.29MB   2530:        _sub_Knl[tag][i].resize(1);
         .          .   2531:
         .    16.50kB   2532:        _jacobian_block_used[tag][i].resize(1);
         .    16.50kB   2533:        _jacobian_block_neighbor_used[tag][i].resize(1);
         .    16.50kB   2534:        _jacobian_block_lower_used[tag][i].resize(1);
         .    16.50kB   2535:        _jacobian_block_nonlocal_used[tag][i].resize(1);
         .          .   2536:      }
         .          .   2537:    }
         .          .   2538:  }
         .          .   2539:}
         .          .   2540:
```

**Comment by lindsayad** [comment]
151 MB to 16.5 kB. Wow so beautiful

File: framework/include/base/Assembly.h [hunk 1]
```diff
@@ -959,16 +959,80 @@ class Assembly
...
+   */
+  void jacobianBlockUsed(TagID tag, unsigned int ivar, unsigned int jvar, bool used)
+  {
+    _jacobian_block_used[tag][ivar][_block_diagonal_matrix ? 0 : jvar] = used;
+  }
+
+  /**
+   * Return a flag to indicate if a particular block is used
+   */
+  char jacobianBlockUsed(TagID tag, unsigned int ivar, unsigned int jvar)
```
Comment (lindsayad) [inline framework/include/base/Assembly.h:977]: Make method `const`

File: framework/include/base/Assembly.h [hunk 2]
```diff
@@ -959,16 +959,80 @@ class Assembly
    */
   DenseMatrix<Number> & jacobianBlock(unsigned int ivar, unsigned int jvar, TagID tag = 0)
   {
-    _jacobian_block_used[tag][ivar][jvar] = 1;
+    jacobianBlockUsed(tag, ivar, jvar, true);
     return _sub_Kee[tag][ivar][_block_diagonal_matrix ? 0 : jvar];
   }
 
+  /**
+   * Flags to check whether or not Jacobian blocks are used
```
Comment (lindsayad) [inline framework/include/base/Assembly.h:967]: I would change this comment to say: `Sets whether or not Jacobian is used to \p used`

File: framework/include/base/Assembly.h [hunk 3]
```diff
@@ -959,16 +959,80 @@ class Assembly
...
+   */
+  void jacobianBlockNeighborUsed(TagID tag, unsigned int ivar, unsigned int jvar, bool used)
+  {
+    _jacobian_block_neighbor_used[tag][ivar][_block_diagonal_matrix ? 0 : jvar] = used;
+  }
+
+  /**
+   * Return a flag to indicate if a particular neighbor jacobian block is used
+   */
+  char jacobianBlockNeighborUsed(TagID tag, unsigned int ivar, unsigned int jvar)
```
Comment (lindsayad) [inline framework/include/base/Assembly.h:993]: Make method `const`

File: framework/include/base/Assembly.h [hunk 4]
```diff
@@ -959,16 +959,80 @@ class Assembly
...
+   */
+  void jacobianBlockLowerUsed(TagID tag, unsigned int ivar, unsigned int jvar, bool used)
+  {
+    _jacobian_block_lower_used[tag][ivar][_block_diagonal_matrix ? 0 : jvar] = used;
+  }
+
+  /**
+   * Return a flag to indicate if a particular lower jacobian block is used
+   */
+  char jacobianBlockLowerUsed(TagID tag, unsigned int ivar, unsigned int jvar)
```
Comment (lindsayad) [inline framework/include/base/Assembly.h:1009]: Make method `const`

File: framework/include/base/Assembly.h [hunk 5]
```diff
@@ -959,16 +959,80 @@ class Assembly
...
+   */
+  void jacobianBlockNonlocalUsed(TagID tag, unsigned int ivar, unsigned int jvar, bool used)
+  {
+    _jacobian_block_nonlocal_used[tag][ivar][_block_diagonal_matrix ? 0 : jvar] = used;
+  }
+
+  /**
+   * Return a flag to indicate if a particular nonlocal jacobian block is used
+   */
+  char jacobianBlockNonlocalUsed(TagID tag, unsigned int ivar, unsigned int jvar)
```
Comment (lindsayad) [inline framework/include/base/Assembly.h:1025]: Make method `const`

**Review by lindsayad — CHANGES_REQUESTED** [review]
All these new methods...is there any reason we would expect to use them outside of `Assembly`? My general inclination is to make methods as limited access as possible until someone gives me a demonstrable reason that they should be more open...so if I was writing this I'd probably start with them `private`

**Comment by fdkong** [comment]
> My general inclination is to make methods as limited access as possible until someone gives me a demonstrable reason that > they should be more open

Good point. I did not see we use these outside of Assembly

File: framework/include/base/Assembly.h (hunk 1 above)
Comment (fdkong) [reply framework/include/base/Assembly.h]: done

File: framework/include/base/Assembly.h (hunk 2 above)
Comment (fdkong) [reply framework/include/base/Assembly.h]: Done

File: framework/include/base/Assembly.h (hunk 4 above)
Comment (fdkong) [reply framework/include/base/Assembly.h]: Done

File: framework/include/base/Assembly.h (hunk 5 above)
Comment (fdkong) [reply framework/include/base/Assembly.h]: Done

File: framework/include/base/Assembly.h (hunk 3 above)
Comment (fdkong) [reply framework/include/base/Assembly.h]: Done

**Review by lindsayad — APPROVED** [review]
Beautiful

**Comment by YaqiWang** [comment]
Thanks Fande.

**Pull request merged**